src/expr/src/expr/expr_jsonb_access.rs

```diff
@@ -0,0 +1,222 @@
+{
+    fn return_type(&self) -> DataType {
+        O::return_type()
+    }
+
+    fn eval(&self, input: &DataChunk) -> crate::Result<ArrayRef> {
+        let Either::Left(path_expr) = &self.path else {
+            unreachable!("optimization for const path not implemented yet");
+        };
+        let path_array = path_expr.eval_checked(input)?;
+        let path_array: &A = path_array.as_ref().into();
+
+        let input_array = self.input.eval_checked(input)?;
+        let input_array: &JsonbArray = input_array.as_ref().into();
+
+        let mut builder = O::new(input.capacity());
+        match input.visibility() {
+            // We could ignore visibility and always evaluate access path for all values, because it
+            // never returns runtime error. But using visibility could save us some clone cost,
+            // unless we adjust [`JsonbArray`] to make sure all clones are on [`Arc`].
+            Some(visibility) => {
+                for ((v, p), visible) in input_array
+                    .iter()
+                    .zip_eq_fast(path_array.iter())
+                    .zip_eq_fast(visibility.iter())
+                {
+                    let r = visible.then(|| self.eval_strict(v, p)).flatten();
+                    builder.output_nullable(r)?;
+                }
+            }
+            None => {
+                for (v, p) in input_array.iter().zip_eq_fast(path_array.iter()) {
+                    builder.output_nullable(self.eval_strict(v, p))?;
+                }
+            }
+        };
+        Ok(std::sync::Arc::new(builder.finish().into()))
+    }
+
+    fn eval_row(&self, input: &OwnedRow) -> crate::Result<Datum> {
+        let Either::Left(path_expr) = &self.path else {
+            unreachable!("optimization for const path not implemented yet");
+        };
+        let p = path_expr.eval_row(input)?;
+        let p = p
+            .as_ref()
+            .map(|p| p.as_scalar_ref_impl().try_into().unwrap());
+
+        let v = self.input.eval_row(input)?;
+        let v = v
+            .as_ref()
+            .map(|v| v.as_scalar_ref_impl().try_into().unwrap());
+
+        let r = self.eval_strict(v, p);
+        Ok(r.and_then(O::output_datum))
+    }
+}
+
+pub fn jsonb_object_field<'a>(v: JsonbRef<'a>, p: &str) -> Option<JsonbRef<'a>> {
+    v.access_object_field(p)
+}
+
+pub fn jsonb_array_element(v: JsonbRef<'_>, p: i32) -> Option<JsonbRef<'_>> {
+    let idx = if p < 0 {
+        let Ok(len) = v.array_len() else {
+            return None;
+        };
+        if ((-p) as usize) > len {
+            return None;
+        } else {
+            len - ((-p) as usize)
+        }
+    } else {
+        p as usize
+    };
+    v.access_array_element(idx)
+}
+
+trait AccessOutput: ArrayBuilder {
+    fn return_type() -> DataType;
+    fn output(&mut self, v: JsonbRef<'_>) -> crate::Result<()>;
+    fn output_datum(v: JsonbRef<'_>) -> Datum;
```

Comment:
What about naming it `to_datum` as its behavior differs from `output` and `output_nullable`? 🤔